File: src/main/java/vip/mystery0/base/springboot/service/RestPlusService.java
package vip.mystery0.base.springboot.service;

import com.fasterxml.jackson.core.type.TypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import vip.mystery0.base.springboot.model.ServiceApiException;
import vip.mystery0.base.springboot.utils.rest.JSON;
import vip.mystery0.base.springboot.utils.rest.RestTemplatePlus;
import vip.mystery0.base.springboot.utils.rest.handler.RestResponseErrorHandler;
import vip.mystery0.base.springboot.utils.rest.interceptor.LoggingClientHttpRequestInterceptor;
import vip.mystery0.tools.java.factory.JsonFactory;
import vip.mystery0.tools.kotlin.model.Response;

/**
 * @author Mystery0
 */
@Service
public class RestPlusService {
    private RestTemplatePlus<Response> restTemplatePlus;

    public RestPlusService() {
        restTemplatePlus = new RestTemplatePlus<>(createRestTemplate(), e -> {
            throw new ServiceApiException(e);
        }, new JSON() {
            @Override
            public <T> T fromJson(String json, Class<T> clazz) {
                return JsonFactory.fromJson(json, clazz);
            }

            @Override
            public <T> T fromJson(String json, TypeReference<T> typeReference) {
                return JsonFactory.fromJson(json, typeReference);
            }

            @Override
            public <T> String toJson(T obj) {
                return JsonFactory.toJson(obj);
            }
        });
    }

    private static RestTemplate createRestTemplate() {
        RestTemplate restTemplate = new RestTemplate();
        restTemplate.getInterceptors().add(new LoggingClientHttpRequestInterceptor());
        restTemplate.setErrorHandler(new RestResponseErrorHandler());
        return restTemplate;
    }

    public void get(String url,
                    Object... uriVariables) {
        restTemplatePlus.get(url, Void.class, uriVariables);
    }

    public <T> T get(String url,
                     Class<T> clazz,
                     Object... uriVariables) {
        return restTemplatePlus.get(url, clazz, uriVariables);
    }

    public <T> T get(String url,
                     TypeReference<T> typeReference,
                     Object... uriVariables) {
        return restTemplatePlus.get(url, typeReference, uriVariables);
    }

    public void post(String url,
                     Object... uriVariables) {
        restTemplatePlus.post(url, null, Void.class, uriVariables);
    }

    public void post(String url,
                     Object request,
                     Object... uriVariables) {
        restTemplatePlus.post(url, request, Void.class, uriVariables);
    }

    public <T> T post(String url,
                      Class<T> clazz,
                      Object... uriVariables) {
        return restTemplatePlus.post(url, null, clazz, uriVariables);
    }

    public <T> T post(String url,
                      Class<T> clazz,
                      Object request,
                      Object... uriVariables) {
        return restTemplatePlus.post(url, request, clazz, uriVariables);
    }

    public <T> T post(String url,
                      TypeReference<T> typeReference,
                      Object... uriVariables) {
        return restTemplatePlus.post(url, null, typeReference, uriVariables);
    }

    public <T> T post(String url,
                      TypeReference<T> typeReference,
                      Object request,
                      Object... uriVariables) {
        return restTemplatePlus.post(url, request, typeReference, uriVariables);
    }

    public void put(String url,
                    Object... uriVariables) {
        restTemplatePlus.put(url, null, Void.class, uriVariables);
    }

    public void put(String url,
                    Object request,
                    Object... uriVariables) {
        restTemplatePlus.put(url, request, Void.class, uriVariables);
    }

    public <T> T put(String url,
                     Class<T> clazz,
                     Object... uriVariables) {
        return restTemplatePlus.put(url, null, clazz, uriVariables);
    }

    public <T> T put(String url,
                     Class<T> clazz,
                     Object request,
                     Object... uriVariables) {
        return restTemplatePlus.put(url, request, clazz, uriVariables);
    }

    public <T> T put(String url,
                     TypeReference<T> typeReference,
                     Object... uriVariables) {
        return restTemplatePlus.put(url, null, typeReference, uriVariables);
    }

    public <T> T put(String url,
                     TypeReference<T> typeReference,
                     Object request,
                     Object... uriVariables) {
        return restTemplatePlus.put(url, request, typeReference, uriVariables);
    }

    public void delete(String url,
                       Object... uriVariables) {
        restTemplatePlus.delete(url, null, Void.class, uriVariables);
    }

    public void delete(String url,
                       Object request,
                       Object... uriVariables) {
        restTemplatePlus.delete(url, request, Void.class, uriVariables);
    }

    public <T> T delete(String url,
                        Class<T> clazz,
                        Object... uriVariables) {
        return restTemplatePlus.delete(url, null, clazz, uriVariables);
    }

    public <T> T delete(String url,
                        Class<T> clazz,
                        Object request,
                        Object... uriVariables) {
        return restTemplatePlus.delete(url, request, clazz, uriVariables);
    }

    public <T> T delete(String url,
                        TypeReference<T> typeReference,
                        Object... uriVariables) {
        return restTemplatePlus.delete(url, null, typeReference, uriVariables);
    }

    public <T> T delete(String url,
                        TypeReference<T> typeReference,
                        Object request,
                        Object... uriVariables) {
        return restTemplatePlus.delete(url, request, typeReference, uriVariables);
    }
}

File: src/main/java/vip/mystery0/base/springboot/service/RestPlusService.kt
package vip.mystery0.base.springboot.service

import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import vip.mystery0.base.springboot.config.BaseProperties
import vip.mystery0.base.springboot.model.ServiceApiException
import vip.mystery0.base.springboot.utils.rest.JSON
import vip.mystery0.base.springboot.utils.rest.RestTemplatePlus
import vip.mystery0.base.springboot.utils.rest.fuse.FuseService
import vip.mystery0.tools.kotlin.factory.fromJson
import vip.mystery0.tools.kotlin.factory.toJson
import vip.mystery0.tools.kotlin.model.Response
import java.lang.reflect.Type

/**
 * @author mystery0
 * Create at 2019/12/25
 */
@Service
class RestPlusService(
    restTemplate: RestTemplate
) {
    private val restTemplatePlus: RestTemplatePlus<Response<*>> = RestTemplatePlus(
        Response::class.java, { _, exception -> throw ServiceApiException(exception) },
        object : JSON {
            override fun <T> fromJson(json: String, clazz: Class<T>): T = json.fromJson(clazz)
            override fun <T> fromJson(json: String, type: Type): T = json.fromJson(type)
            override fun <T> toJson(obj: T): String = obj.toJson()
        },
        restTemplate
    )

    fun <T> get(
        url: String,
        type: Type = Void::class.java
    ): T? = restTemplatePlus.get(url, type)

    fun <T> getForEntity(
        url: String,
        type: Type = Void::class.java
    ): ResponseEntity<T>? = restTemplatePlus.getForEntity(url, type)

    fun <T> post(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): T? = restTemplatePlus.post(url, request, type)

    fun <T> postForEntity(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): ResponseEntity<T>? = restTemplatePlus.postForEntity(url, request, type)

    fun <T> put(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): T? = restTemplatePlus.put(url, request, type)

    fun <T> putForEntity(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): ResponseEntity<T>? = restTemplatePlus.putForEntity(url, request, type)

    fun <T> delete(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): T? = restTemplatePlus.delete(url, request, type)

    fun <T> deleteForEntity(
        url: String,
        request: Any? = null,
        type: Type = Void::class.java
    ): ResponseEntity<T>? = restTemplatePlus.deleteForEntity(url, request, type)
}